sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -112,16 +117,22 @@ RT::PiKernel ProgramManager::getOrCreateKernel(OSModuleHandle M,
   }
   RT::PiProgram Program = getBuiltOpenCLProgram(M, Context);
   std::shared_ptr<context_impl> Ctx = getSyclObjImpl(Context);
+
   std::map<RT::PiProgram, std::map<string_class, RT::PiKernel>> &CachedKernels =
       Ctx->getCachedKernels();
-  std::map<string_class, RT::PiKernel> &KernelsCache = CachedKernels[Program];
-  RT::PiKernel &Kernel = KernelsCache[KernelName];
-  if (!Kernel) {
-    PI_CALL(RT::piKernelCreate, Program, KernelName.c_str(), &Kernel);
-    // TODO need some user-friendly error/exception
-    // instead of currently obscure one
+
+  {
+    std::lock_guard<std::mutex> CacheLock(Ctx->getProgramKernelCacheMutex());
+    std::map<string_class, RT::PiKernel> &KernelsCache = CachedKernels[Program];
+    RT::PiKernel &Kernel = KernelsCache[KernelName];
+    if (!Kernel) {
+      PI_CALL(RT::piKernelCreate, Program, KernelName.c_str(), &Kernel);
+      // TODO need some user-friendly error/exception
+      // instead of currently obscure one
+    }
```

Comment:
Same here, we should be able to different kernels in parallel without any issues. I suggest reducing the critical section to contain code only related to programs/kernels cache operations

Reply:
Done.
BTW, with split critical section there may be spurious extra call to piKernelCreate. The excess kernel will be then released if there is already kernel in cache. This is example scenario:
1. Thread1 and thread2 enter ProgramManager::getOrCreateKernel with the same kernel name
2. Thread1 locks cache mutex and sees there is no kernel in cache. Thread2 awaits for mutex release.
3. Thread1 releases mutex and enters piKernelCreate, thread2 locks mutext sees there is no kernel.
4. Thread1 returns from piKernelCreate, thread2 releases mutex and enters piKernelCreate.
5. Thread1 locks cache mutex and commits changes to the cache whilst releasing the mutex 
6. Thread2 locks cache mutex and sees that there is already a kernel in cache. Thread2 releases recently created kernel.

Reply:
Added TODOs on possible improvement with elimination of this scenario